### ISSUE TITLE
Support remote execution with rules_nixpkgs

### ISSUE THREAD

**Issue by aherrmann** [tweag/rules_nixpkgs#180]
**Is your feature request related to a problem? Please describe.**
Bazel supports [remote execution](https://docs.bazel.build/versions/main/remote-execution.html) through the [remote execution protocol](https://github.com/bazelbuild/remote-apis/blob/main/build/bazel/remote/execution/v2/remote_execution.proto). This protocol manages inputs and outputs required by and generated by Bazel build actions, i.e. actions defined by [regular Bazel rules](https://docs.bazel.build/versions/main/skylark/rules.html).

However, `rules_nixpkgs` defines [repository rules](https://docs.bazel.build/versions/main/skylark/repository_rules.html) that invoke `nix-build` during Bazel's loading phase. The Nix package manager will then realize Nix store paths (typically under `/nix/store/...`) and generate symlinks into Bazel's execution root. These Nix store paths are outside of Bazel's control and the remote execution protocol does not ensure that these store paths are also realized on the remote execution nodes.

Remote execution actions that depend on Nix store paths will fail if the required Nix store paths are not realized on the remote execution nodes.

**Describe the solution you'd like**
We need some solution to ensure that Nix store paths that are required for Bazel build actions exist on the remote execution nodes that these actions may be run on.

Some possible approaches:
* Run a script before each Bazel invocation (e.g. by defining a custom [`tools/bazel`](https://github.com/bazelbuild/bazel/issues/5788)) that builds all required Nix store paths on the remote execution nodes.
* Remote execution nodes could share a Nix store via network fs.
* `nixpkgs_package` could set [`remotable = True`](https://docs.bazel.build/versions/5.0.0/skylark/lib/globals.html#repository_rule(implementation,%20attrs,%20local,%20environ,%20configure,%20remotable,%20doc)) to execute the `nix-build` command on the remote execution nodes. ([feature announcement](https://groups.google.com/g/bazel-discuss/c/xm0wgsCzTG0/m/d20QwsbrBQAJ), [commit](https://github.com/bazelbuild/bazel/commit/06a26c37126e3eefabd28c24a4e35059b5b136bc), [flag](https://docs.bazel.build/versions/main/command-line-reference.html#flag--experimental_repo_remote_exec))
* Remote execution [could be configured](https://www.buildbuddy.io/docs/rbe-platforms#using-a-custom-docker-image) to run actions in Docker images that contain the needed Nix store paths, perhaps using [Nixery](https://github.com/tazjin/nixery).
* Place the Nix store inside Bazel's execroot (perhaps using [managed directories](https://docs.bazel.build/versions/main/skylark/lib/globals.html#workspace)) and add a [sandbox mount pair](https://docs.bazel.build/versions/main/command-line-reference.html#flag--sandbox_add_mount_pair) (should support [relative paths](https://github.com/bazelbuild/bazel/blob/11f7d8040fadc595589ee264561606dc2a83685d/src/main/java/com/google/devtools/build/lib/sandbox/LinuxSandboxedSpawnRunner.java#L299)) to mount it under `/nix/store` inside the sandbox.

cc @YorikSar @AleksanderGondek @r2r-dev

**Comment by Jonpez2** [comment]
If we could find a way to run every action, whether local or remote, under something equivalent to nix-shell --pure -p <the-nix-packages-you-told-me-you-depend-on>, then that would be best, right?  Then it wouldn't matter whether you were local or remote, and you'd never get inhermeticity.
I'm certain there's a flake approach here too - maybe before every action we emit a flake.nix with appropriate dependencies and nix run that?

**Comment by Jonpez2** [comment]
Does this seem credible?

**Comment by aherrmann** [comment]
@Jonpez2 Perhaps yes, though this may still require additional metadata: To achieve fine granularity the set `[the-nix-packages-you-told-me-you-depend-on]` would need to be minimal for each action. So, it's not one global Nix environment for the entire build, but precise Nix dependencies for each build action. The way we integrate nix with Bazel through rules_nixpkgs the Nix store paths are not provided through a nix shell environment, but instead through symlinks into the nix store (with gc roots), i.e. rules_nixpkgs invokes `nix-build`.

**Comment by AleksanderGondek** [comment]
@aherrman @Jonpez2
Apologies for I have been living on a quite rules_nixpkgs-distant planet for a while :D 

In principle, every Bazel action should be hermetic and pure - therefore it stands to reason that running it within a very narrowly defined nix-shell achieves that goal and I would love to be able to do just that.

However, due to my experiences of running rules_nixpkgs and Bazel in tandem (which makes me also biased and blind to “newcomer” perspective) I see one major, troublesome aspect of proposed approach:

Interoperability with existing Bazel ecosystem / rules.

I have not fully thought this through, but it seems to be that unless the change would be placed in Bazel code itself (sic!), all of existing rules would need to change to be able to act on inputs delivered from nixpkgs - example `cc_library` would need to recognize inputs provided by nix package manager and act on them differently then the ones from Bazel itself). Great deal of composability is sacrificed.

**Comment by Jonpez2** [comment]
Yeah agreed. I think we need to lobby someone in bazel land to figure out
how to do this natively. It probably will require internal changes…

**Comment by aherrmann** [comment]
> I have not fully thought this through, but it seems to be that unless the change would be placed in Bazel code itself (sic!), all of existing rules would need to change to be able to act on inputs delivered from nixpkgs

Correct, the proposed "If we could find a way to run every action, whether local or remote, under something equivalent to nix-shell [...]" would, I think, require a change to Bazel itself. An alternative that we've been thinking about, if we're to modify parts of Bazel anyway, is to tackle this at the remote execution protocol level. That protocol currently has no notion of a "system root" or other kinds of per-action system dependencies. If the protocol could express such constraints per-action in a generic way, then the remote side could ensure that the constraints are resolved before the action is executed. E.g. that `nix-build` is run to provide the needed Nix store paths.

Side note, this is, not directly but still somewhat tangentially, related to https://github.com/bazelbuild/bazel/issues/6994.

**Comment by Jonpez2** [comment]
FWIW, if we're thinking of something remote-execution-specific, BuildBarn has the following two interesting (and I think related) issues:
https://github.com/buildbarn/bb-remote-execution/issues/40
https://github.com/buildbarn/bb-remote-execution/issues/23

**Comment by Jonpez2** [comment]
@AleksanderGondek - re the Remote Assets API - would that mean something like doing a nix query to find the transitive closure of required /nix/store roots, and then having the remote worker unpack into a /nix/store that looks exactly like that?  And then having starlark code which figures out resolved paths within /nix/store, and executing with that?  That seems a bit entwined and fragile to me...

**Comment by aherrmann** [comment]
> There is another way to move forward, which I feel is a bit more lean and less disruptive towards overall Bazel build model.
> 
> Bazel has [Remote Assets API](https://github.com/bazelbuild/remote-apis/blob/main/build/bazel/remote/asset/v1/remote_asset.proto) that can be extended to provision `/nix/store`-bound artifacts.

What I remember of this approach from the last attempt was that it only worked when the remote execution system could still access the needed nix files on the host directly. That's usually not true, e.g. when a developer issues a build on their development machine and the remote execution service runs on a cloud platform or somewhere else on a remote machine. If that limitation could be fixed then this could indeed be a viable option.

**Comment by Jonpez2** [comment]
I really think the only safe option is a nix-shell or nix run flake style wrapper.
Do all actions happen within the context of some particular posix toolchain by any chance?  e.g. do they all invoke via a bash selected from the toolchain, or something like that?  Is there any way we could hook that?

**Comment by Jonpez2** [comment]
Another point that occurred to me this morning - I think we need to define a Provider which adds nix-specific data for a rule: i.e. which nix packages are require to build the rule, and which are required to run the rule.
Then maybe we figure out how to plug in a host-package-manager manager into bazel itself, which consumes data out of such a provider (or the transitive closure of the providers collected from the deps?) , and pre-sets-up the host env for the rule's action executions.
I say this because some rules may run on remote host a), resolve a /nix/store path, and then hand it over to another rule which proceeds to run on remote host b) which hasn't got that resolved.  So we need to communicate requirements between rules I think.

**Comment by Jonpez2** [comment]
Does all of this apply to guix as well? Ie could we make it a bit more palatable to bazel by making it apply to at least 2 mainstream package managers?
Do you in tweag have interest in a call to discuss possible ways forward on this? I would be excited to give some context on my usecase.

**Comment by Jonpez2** [comment]
Yeah I really don’t think this should happen via the remote execution api. On nix, this can and therefore should work exactly the same across remote and local, no?

**Comment by aherrmann** [comment]
I'm currently a bit overloaded on other work and OOO due to an accident. I haven't managed to catch-up on the discussion above, yet. I haven't forgotten it and I intend to contribute further, just letting you know to manage expectations.

**Comment by Jonpez2** [comment]
I hope everyone involved is ok!

**Comment by Jonpez2** [comment]
Hello again!  No-op sort of an update: I've been having a dig around in the bazel codebase to try to figure out some kind of a way forward here, but I haven't come up with anything particularly useful.  I am thinking that we would want to add a spawn strategy which wraps the DynamicSpawnStrategy and somehow picks out the transitive closure of relevant NixToolProvider (i.e. the thing I was trying to describe in https://github.com/tweag/rules_nixpkgs/issues/180#issuecomment-1285735716).  Then "all it would need to do" is, in the exec() function, before delegating to DynamicSpawnStrategy, prefix the spawn's command line with 'nix-shell -i [whatever] run -- " or something?
There's a bit of handwaving in there :)

**Comment by Jonpez2** [comment]
Gentle ping on this one
FWIW, here is a googel groups thread on the subject - https://groups.google.com/g/bazel-discuss/c/kqv-EHhApbY

**Comment by aherrmann** [comment]
Sorry for the long silence on this one. I've been quite busy lately.

> Do all actions happen within the context of some particular posix toolchain by any chance?

@Jonpez2 That is one of the difficulties, at least for a generic solution on the level of rules_nixpkgs, i.e. from the library author's perspective: We don't what kinds of targets users will import from Nix, and we don't know how users will use these targets. They could import toolchains, libraries, build tools, a Docker base image, etc. So, actions that use Nix provided files could be any action really. See also what @uri-canva points out in https://github.com/tweag/rules_nixpkgs/issues/180#issuecomment-1284666194.

> Note there's some extra considerations if the host machine is a different platform than the remote executor. The common way of using rules_nixpkgs right now is to let nix detect the host machine platform, instead of having bazel pass that platform information in (which it can't even do for repository rules anyway, it can only do it for actions).

@uri-canva That's correct. I think handling this correctly is possible. The `nixpkgs_package` import could explicly set the `system` argument to `nixpkgs` and the generated toolchain could set the correct exec constraints. The assumption that host equals exec may be hard-coded in some places that would need fixing. But, I think there's no strong technical reason for that assumption, mostly just historical reasons. It's clearly related to this ticket here, but it's a separate issue. @uri-canva do you want to go ahead and open a feature request for it?

> No that's definitely the approach that would be the most compatible with bazel: defining toolchains using the bazel APIs in a way such that bazel doesn't need to know anything about nix.

@uri-canva The trouble is that Nix provided dependencies can be used in actions that don't have a natural associated toolchain. E.g. a `genrule` or a custom action.

> For example if we define toolchains with binaries that are nix-shell wrappers

@uri-canva Two things that make this problematic:
1. Nix-shell wrappers defer the Nix-evaluation to the execution time. This means we have shifted the problem from having to ship the Nix store paths to the remote side to having to ship the Nix expressions to the remote side. I.e. the Nix sources that the Nix shell wrapper loads and evaluates have to exist on the remote executor under the correct path.
2. Nix evaluation adds overhead, sometimes quite considerable one. The nice thing about the way rules_nixpkgs does it right now is tha this evaluation happens only once per import when fetching the `nixpkgs_package`. With Nix shell wrappers this would happen every time. This can be problematic. I've worked on projects in the past where this overhead made Nix shell wrappers infeasible for certain tools, e.g. `gcc`.

> One approach that we've spiked is using nix to build portable binaries that are self contained, using `buildFHSUserEnv`. That's indeed a nice solution. But, as you point out has it's costs. At the level of rules_nixpkgs I'd prefer a more generic solution, if possible. I think it's worth pointing out the difference in the user and the library author perspective here: As a user with a concrete use-case in mind one can craft a dedicated solution that fits well with the given infrastructure and codebase, and one can adjust the own project around the constraints that the chosen solution implies. As the library authors of rules_nixpkgs we should strive to find a solution that doesn't unduly restrict or dictate the setup on the user. In practice we may have to be pragmatic here and there and impose some restrictions to arrive at a feasible solution, but we should try not to be overly restrictive. I feel like forcing users to turn every Nix imported package into a self-contained, relocatable package is probably too restrictive. That said, if it works at a use-site of rules_nixpkgs and fits well into a given project, sure why not.

> Another point that occurred to me this morning - I think we need to define a Provider which adds nix-specific data for a rule

@Jonpez2 Keep in mind that rules_nixpkgs implements repository rules and these cannot return providers. Providers only exist at the level of regular rules. That said, if this only about transmitting metadata to dedicated rules, aspects, or other tools through some additional mechanism. Yes, that can be a viable route, rules_nixpkgs could auto-generate such metadata targets. Take a look at the good work done by @AleksanderGondek and @r2r-dev on https://github.com/tweag/nix_gazelle_extension - it generates dedicated metadata targets to transmit Nix metadata into Gazel. I think it's somewhat different from what you're point at here, but may still be a good point of reference.

> Note that changes to the remote execution APIs are a bit more complex to get through, since there's several implementations of it. See https://github.com/bazelbuild/remote-apis#api-community.

@uri-canva Absolutely, I understand. The observation that makes me consider this route is that rules_nixpkgs is not alone with the problem of having to transmit information about system dependencies or other kinds of ambience. Projects that use distribution package managers could also benefit from the ability to collect system dependencies across targets and ship this metadata to the remote side. Indeed, to extend the protocol (if needed) to better support rules_nixpkgs would have much higher chances of success if it could also benefit other non-Nix related use-cases.

**Comment by aherrmann** [comment]
@Jonpez2 Thanks for the mailing list thread. I'll take a look. I saw Alex Eagle mentioned the community day session there. I've been meaning to share my notes here. I'll try to do so now:

[BazelCon Community Day](https://sites.google.com/engflow.com/bazelcon-2022-community-day), which was held one day before BazelCon, included unconference style sessions, and we had one session on remote execution and system dependencies and Nix in particular.

Many voices suggested to find a way to track the Nix store paths with Bazel explicitly, i.e. somehow store Nix store paths under Bazel's output-base and mark them as regular dependencies such that the remote execution system would automatically push them to the remote side. The problem here is that nixpkgs assumes stable absolute paths and Bazel's action execution evironment does not provide these. So, this would require some form of path remapping, to map Nix store paths fetched into Bazel’s output base to some stable absolute path. E.g. some form of file system abstraction.

A promising suggestion came from @illicitonion. Namely, mark Nix imported targets with special platform properties that define which Nix derivation is required to run the action. Then, add a feature to Bazel to accumulate platform properties across transitive dependencies such that the transitive closure of required Nix derivations is communicated to the remote side for each action. Finally, extend the remote executor to parse these platform properties and ensure that the required Nix derivations are instantiated before running the action.

The underlying observation is that there are really two problems here:
1. What Nix derivations does a given action depend on.
2. How do make sure that the corresponding Nix store paths are instantiated at the remote side.

The marking and transitive accumulating of the platform properties achieves 1. Conveniently, platform properties are already shipped to the remote side, so that this doesn't require extension of the RBE protocol.
2. then has to be addressed on the remote executor side by reading these platform properties and instantiating the corresponding store paths.

A concern that was brought up is that we may want to cut the Nix dependencies at some point, or distinguish runtime and build time deps.

**Comment by Jonpez2** [comment]
@aherrmann would you be open to having a call on this? Is there a way for me to contact you directly to set one up please?

Thank you!

**Comment by k1nkreet** [comment]
@olebedev I'm experimenting with this issue and I'm trying two ideas both has a lots in common with what you are proposing. I will try to describe what am I trying and how it could be connected with your ideas, and maybe we'll be able to figure out something working from that.

So the first idea I'm playing with is to make a nix-closure of package a Bazel target like you are suggesting with docker image. I was thinking about building a tarball as a target which would be carried by Bazel as usual dependency. There is several problems with it: 

  1. It would bring a lot of performance penalties and a lot of space waste
  2. The targets dependent on it should be aware about how to use this tarball, or there should be something in the middle unpacking it and generate usable targets to depend on

The second idea is to insert a dummy target like a genrule running nix-build into the dependency chain, so every target dependent on `nixpkgs_package` outputs would depend on this target and make it tagged as `no-cache`l. In this case this target will be run every time by remote executor creating items in the nix-store if they are not exists or being a no-op otherwise.

Regarding the approach you've described I have couple questions: are you proposing using this docker image as an output of `nixpkgs_package` instead the filegroups for example. How would it's dependencies be aware of how to utilize it?  When you are proposing to ensure this genrule would be run by the remote executors why would you need to produce this docker image?

**Comment by olebedev** [comment]
> are you proposing using this docker image as an output of nixpkgs_package instead the filegroups for example.

@k1nkreet, not really, sorry. I should have been clearer about the proposal, the example with the docker image is an example of building a Nix package using the `genrule` but not really about using this docker image to solve the original problem anyhow. It could be basically any Nix package, and more appropriate example would be using my previous example but implementing it as a `genrule`. Something like this:

```bzl
# NOTE: This is a pseudo-code snippet to demonstrate the idea
genrule(
  name = "libxxf86vm",
  srcs = ["default.nix", "//pkgs:default.nix"], # properly declared dependencies, using glob, filegroups etc
  outs = ["nixpkgs-libxxf86vm"],
  cmd = """
nix-build $(location default.nix) -o "$@" -A "xorg.libXxf86vm"
""",
  tags=["requires-network"],
)
```

> I was thinking about building a tarball as a target which would be carried by Bazel as usual dependency. 

Yeah, we have discussed it today with my colleagues and we ended up with the exact same conclusion - the tarballs or docker images per Nix package/closure will bring a huge data transfer overhead and destroy the idea of incrementality/granularity per Nix package.

But the general idea of having a hook that is being executed on a particular remote executor doesn't sound invalid to me but needs elaboration.

For example, instead of having a tarball per Nix package/closure and making sure we have extracted the tarball properly, we would need to have some sort of a hook that is being executed straight before the main action command, right? If this is the case, I think, it would be way more efficient to pass over to the executor a set of the Bazel inputs for that particular Nix package and invoke something based on the `nix-build` CLI  instead of invoking `tar xf ...`. By that, we remove the performance penalties you have been talking about, what do you think?

> In this case this target will be run every time by remote executor creating items in the nix-store if they are not exists or being a no-op otherwise.

Are you sure that would work as expected, will the Bazel scheduler pass through all the inputs that the Nix package depends on to be able to build it correctly? Meaning, would Bazel care about passing additional inputs set for the prerequisite target to be executed before the main execution. From my understanding of how the BRE works, it won't be working that way. I would be happy to be wrong and this would be a really elegant solution for the whole problem if this is the case.

Let me know what do you think.

**Comment by aherrmann** [comment]
I raised this issue in today’s [Remote Execution API Working Group monthly meeting](https://docs.google.com/document/d/1EtQMTn-7sKFMTxIMlb0oDGpvGCMAuzphVcfx58GWuEM/edit?usp=sharing). Here is the outcome of that discussion:

- The RBE protocol already effectively has two notions of shipping directory trees to the remote execution side: The [input root of the build action](https://github.com/bazelbuild/remote-apis/blob/3a21deee813d0b98aaeef9737c720e509e10dc8b/build/bazel/remote/execution/v2/remote_execution.proto#L463-L470) and a Docker image through platform properties, which is arguably also a directory tree. It would be preferable to not have to introduce yet another notion of directory tree input into the protocol to support the Nix use-case.
    - A feature that could remap input paths to arbitrary paths in the remote execution environment would be very welcome. This could be a way to support Nix builds. If we import Nix store paths as regular build dependencies under Bazel’s output root and then remap them to the correct `/nix/store` path on the remote side.
    - Relatedly, BuildBarn already has a feature [to chroot into the input root](https://github.com/buildbarn/bb-remote-execution/issues/23) that @Jonpez2 had also mentioned before. Perhaps this feature could already be used to support Nix+Bazel remote builds today. BuildStream also uses this feature.
    - There is a [related thread on the mailing list](https://groups.google.com/g/remote-execution-apis/c/1rzZxfY34vI/m/-gXBsJ9qAAAJ) about including a container image in the input root.
- One concern that was raised was about cross-platform projects, e.g. if a user initiates a Bazel remote build on their MacOS host that will run on a Linux remote execution cluster. This relates to [the concern](https://github.com/tweag/rules_nixpkgs/issues/180#issuecomment-1284656149) raised by @uri-canva above.
    - In principle Nix and nixpkgs should provide the necessary features to support this use-case through the `system` parameter to `nixpkgs` and the Nix remote cache or Nix remote builds if necessary.
- There was some support for [the idea to list the required Nix derivations in per-action platform properties](https://github.com/tweag/rules_nixpkgs/issues/180#issuecomment-1329081409) and have the remote executor ensure that they are present before executing the action.
- Another suggestion was to provide a test wrapper, and build tool wrapper, that ensures that the required Nix store paths are realized before proceeding with the build action or test. This is related to [the wrapper idea](https://github.com/tweag/rules_nixpkgs/issues/180#issuecomment-1284251637) discussed earlier on this issue.
    - As mentioned above, this could potentially work for executables, though care would need to be taken that Nix evaluation time is not too costly. But, this would not work with non-executable dependencies like shared libraries.

**Comment by aherrmann** [comment]
> > The `nixpkgs_package` import could explicly set the `system` argument to `nixpkgs` and the generated toolchain could set the correct exec constraints.
> 
> @aherrmann I did that in my static toolchain prototype by passing `nixopts = ["--argstr", "system", "x86_64-linux"]` and similar to `nixpkgs_package` and it worked well. I will open separate issues if I see any instances of the host equals exec assumption.

@uri-canva Thanks for testing! I'm glad to hear it works!

> > The trouble is that Nix provided dependencies can be used in actions that don't have a natural associated toolchain. E.g. a `genrule` or a custom action.
> 
> @aherrmann Bazel is going pretty hard on toolchains, so I think we can depend on them in the general case.

I fear I'll have to disagree on this. Yes, there is a strong push for platforms & toolchains, but, mostly for rules were the complexity is warranted, e.g. language rules like `LANG_library`, `LANG_binary`, etc. However, for simpler, or ad-hoc tasks a `genrule` or similar is a common choice. If you look at [Alex Eagle's talk at Bazel Community Day](https://sites.google.com/engflow.com/bazelcon-2022-community-day?pli=1) you'll even hear it recommended to use `run_binary` for simple cases.

> Yes, genrules and other very specific actions might require special case support, like configuring the right --shell_executable when launching bazel, or ensuring that the remote execution environment and the local execution environment have the exact same shell available.

That is a global approach though, it can work for the standard shell environment, but it's not appropriate for more special purpose tools that are only required by few rules. Because making them part of the global environment would invalidate all targets on change.

Where toolchains are appropriate it makes sense to use them. But, I think we'll always have to deal with rules or inputs that don't have a corresponding toolchain. E.g. tools for ad-hoc code generation steps, or tools to support test-cases, etc.

> > Two things that make this problematic:
> 
> @aherrmann Agree. We have thought / prototyped down this path more and it introduces a lot of pain points, especially if you want to avoid leaking nix specifics across the remote execution interface.
>
> > I feel like forcing users to turn every Nix imported package into a self-contained, relocatable package is probably too restrictive.
> 
> @aherrmann Yes, I was hoping there could be a more general way of making this work, like using `pkgsStatic` from nixpkgs, but I found it didn't quite work well in practice. I'm now looking at other ways of providing relocatable packages.

Thank you for looking into these and reporting back on your experiences! Please keep us in the loop.

**Comment by sluongng** [comment]
> There was some support for https://github.com/tweag/rules_nixpkgs/issues/180#issuecomment-1329081409 and have the remote executor ensure that they are present before executing the action.

I think realistically, this would be the cheapest and easiest-to-implement solution.

Most RBE vendor today has already built some support for Container Image for special actions.
I can see how similar support could be provided for Nix.

---

Long-term wise, I think Remote API is not the right place to embed Nix logic. As folks from the meeting have highlighted, the API currently supports a custom root tree for the action, which is a very coarse, but sufficient level of support encompassing Nix use cases.

I think it would be nice if Bazel itself could provide some sort of customization to actions. Perhaps some sort of hooks to prepare/clean-up the execution sandbox prior to each action run. Which could be transparently turned into a wrapper running commands before/after each action.

**Comment by Jonpez2** [comment]
Re static binaries - one thing we should keep in mind is generation of
container images from normal bazel rules which depend on nix derivations -
we would, I assume, want to use nix-native code to do all the appropriate
symlinking, env vars, and whatever else is required? So I think an approach
that cuts out nix metadata would preclude this?

A motivating example might be a library which wraps up invocations of
kubectl with the gke-gcloud-auth-plugin all correctly setup. Then we take
that as a dependency and invoke a rules_docker (or maybe better a
rules_nix_containers!) to build a container image. I think we could make
that work only if we retain nix metadata, right? Maybe the same question
applies to a java_binary that takes that lib as an input too.

Thank you for all of the work on this, it’s so exciting to see this move
forward. It will be such a leap forward for trivial bazel hermeticity.

**Comment by olebedev** [comment]
> but that gets a bit tricky because you'd have to verify the container is available in the registry on every build and push it if it's not, or similar stuff.

@uri-canva, this look similar to the existing problem with verifying if Nix artifacts have been populated to the /nix/store and perform some actions if they're not. 

> It's also possible to make the interface just the container runtime, and inline a container in the build

Sorry for potentially silly question, I am not super familiar with the Bazel's internals and the Bazel's remote execution API, please bear with me. @uri-canva, if we go that path and build containers inline, do we really need that roundtrip with pushing these containers to registry and then pull them back in on each remote executor machine, meaning, would it be more appropriate to remove this registry part and let BRE API work out how these inline built images delivered to the remote execution machine?

**Comment by aherrmann** [comment]
> There are 3 ways of getting binaries into your bazel build:
> [...]
> In the context of rules_nixpkgs, I believe we're only interested in 3.

@uri-canva I'd distinguish another way in this case: "Hard-coding an absolute path to an external binary".
That is what rules_nixpkgs effectively does, either through a symlink (`bin/...`) that is then tracked as a `File` in Bazel, or through a string, e.g. in the cc toolchain configuration. (Similarly, on Windows, one may [create shims](https://github.com/tweag/rules_sh/blob/25e6d80250aa130b112976c47635e686d99f984c/sh/import.bzl#L21).)

> At the moment the targets those nixpkgs_package rules generate are not correct, as they don't declare all their runtime dependencies, the generated targets just glob the files.

Yes, in the sense that they don't expose all needed runtime dependencies as artifacts to Bazel.

Unfortunately, there are technical reasons why it's not always feasible to expose all needed files as artifacts to Bazel.
E.g. tracking a full Python toolchain, even just as symlinks into the Nix store, [will quickly use up too much disk](https://github.com/tweag/rules_nixpkgs/pull/103). For large sets of inputs there is also considerable overhead in input hashing, [see here](https://github.com/bazelbuild/bazel/pull/13879#issuecomment-1263983209).

> because the absolute paths baked into the files would lead to them referencing the files outside of the bazel sandbox.

Correct.

> Since the targets aren't defined correctly from bazel's point of view, they do not work with remote execution.

Yes, put this way, this also hold for all other kinds of global dependencies, e.g. `/usr/bin/bash`, or `/usr/lib64/ld-linux-x86-64.so.2`. Of course the difference being that these are commonly installed on the remote executors as well.

> In principle I like the idea of defining correct targets very much, as it's very idiomatic from the bazel point of view, which helps people who use bazel but might not be familiar with nix work with it and debug it. 

In principle I agree. As mentioned above, unfortunately there are technical limitations in Bazel that make this not always feasible. For the case of "correct" meaning all runtime dependencies explicitly tracked as files by Bazel.

**Comment by k1nkreet** [comment]
I want to share the experience I got when I was experimenting with this hacky little idea I've described above, in case someone would find it useful or maybe would want to experiment with it. Briefly, the idea was to create an ordinary rule which runs `nix-build` and specifies the outputs which could be used as inputs by ordinary Bazel rules like `cc_library` for library objects and headers or toolchains for binary outputs. This `nix-build` action should be marked as `no-cache` and `no-remote` and should be instantiated for every separate group of outputs  which is could be used as a separate inputs for other rules. This approach tries to insert nix-build execution into the dependency chain securing the existence of store paths relying on the Nix cache which allow us call to build the same derivation multiple times without a huge penalty. It also implies all remote builders are sharing the same Nix store. 

For example if I want to build `zlib` C-library it should end up as  two `nix_build` rule instantiations: one to produce shared objects and one to produce headers. Those outputs could be used as a `srcs` and `hdrs` of `cc_library` respectively. It turned out to be working with one adjustment: these rules have to specify all the outputs before-hand, and I've decided I can use `nixpkgs_package` repository_rule for this. It can run `nix-build` as it does now, collect the outputs and produce proper `nix_build` rules.

Nevertheless, there is several problems which forces me to stop this experiment:

  * Nix gc-roots cannot reliably be kept alive as long as they may be needed. If the `nix-buil`d action creates the Nix gc-root in its execroot, then it will be deleted by Bazel after executing the action. Furthermore, If the remote execution nodes have read-only access to the shared Nix store then they cannot communicate gc-roots to the node managing the Nix store.
  * The Nix store has to be shared, but concurrent writes from multiple Bazel workers to the same Nix store database is problematic. So, this may also require Nix remote builds
  * It's unclear how to deal with runtime dependencies

**Comment by sluongng** [comment]
> This approach tries to insert nix-build execution into the dependency chain securing the existence of store paths relying on the Nix cache which allow us call to build the same derivation multiple times without a huge penalty. It also implies all remote builders are sharing the same Nix store.

This means the store would be uploaded from your local laptop to Remote Cache CAS, and then fetched by remote executor workers as subsequent action inputs right?
Does that mean there is a potential where the cache may differ between different users?

**Comment by olebedev** [comment]
> No since this rule is marked as no-cache it will never upload anything into Remote CAS. Remote workers will be forced to re-run it which will secure existence of nix store paths on the remote side

@k1nkreet, wouldn't it be executed only on the host (the build host machine) as you assign `no-remote` tag to the target?

**Comment by k1nkreet** [comment]
@olebedev Oh sorry, this was a typo, I was going to write `no-sandbox`. Thanks for noticing!

**Comment by aherrmann** [comment]
@uri-canva Thank you for the update!

> The most common bazel remote execution setup doesn't just assume exec to be linux FHS with glibc, but the target platform as well.

Yes, indeed, that's something that could and probably should be considered more carefully in rules_nixpkgs. We currently mostly only set the `exec_constraints` to include `support_nix`, but, as you rightly point out, this is often also true for  `target_constraints`. For reference the point came up in a past discussion [here](https://github.com/tweag/rules_nixpkgs/issues/109#issuecomment-601158963):
> A `py_binary` built with a `nixpkgs` provided Python toolchain, will require the Python interpreter in the Nix store path at runtime, so the `nixpkgs` constraint should go into `target_compatible_with`. Currently, it is only set on `exec_compatible_with`.

That said, there are some ways to make rules_nixpkgs produced artifacts deployable on non-Nix systems, see [here](https://github.com/tweag/rules_nixpkgs/issues/291) for some pointers. But, upstream support for this in nixpkgs would be great.

> I think a nix system is such a platform, but a nix system with a specific set of packages in the nix store is not.

Yes, I think platform constraints are too coarse to go into the level of individual packages. Currently rules_nixpkgs just uses [`support_nix`](https://github.com/tweag/rules_nixpkgs/blob/726e2b74a59047b2f9eb37c1c4072e2c74465792/core/constraints/BUILD.bazel#L6) (mostly just in `exec_constraints`) to indicate whether Nix is installed on the platform. In principle there is also [`@platforms//os:nixos`](https://github.com/bazelbuild/platforms/blob/3fbc687756043fb58a407c2ea8c944bc2fe1d922/os/BUILD#L94), however, that makes NixOS a different OS than Linux, which doesn't work too well for projects that want to build on a regular Linux system (e.g. Ubuntu) with Nix installed as a package manager.

**Comment by sluongng** [comment]
@uri-canva I think your analysis is on point here.

> Given a codebase using nixpkgs_package and nixpkgs_*_configure rules for host builds, there could be an automated way to produce the container image for remote execution, and all the toolchains required to use it.

Keep in mind, container image support does not come with Bazel by default.
This is a "hack" outside of Bazel, provided by various different RBE implementation. By including a container image via `exec_properties` as part of the Execute rpc request.

This routes back to my original point earlier: perhaps, what it would take for Nix to work with RBE is simply another similar hack on RBE side to enable nix to work. Here is what I have in my head:

1. Have a local process, pre-build, to upload needed Nix files to the remote cache under 1 directory(tree):
   ```
   shell.nix
   flake.nix
   flake.lock
   ```

2. Include the digest of the tree above as part of `exec_properties` of Bazel action ([example](https://www.buildbuddy.io/docs/rbe-platforms#using-a-custom-docker-image)), telling RBE platform to select builders matching this requirement.
   ```
   exec_properties = {
        "OSFamily": "Linux",
        "nix-tree": "SHA256:aaaaaaaaaabbbbbbbbbbbccccccccccccc",
    },
   ```

3. Special RBE builders fetch the tree digest and run `nix shell` to prepare working directory / environment variables before executing an action. These special builders could have their own custom way of integrating with a Nix Remote Cache or even a Nix Remote Build setup.

Would something like this works or is there some use cases that I am missing here?

**Comment by aherrmann** [comment]
@uri-canva Thank you for sharing your progress on this issue and this classification. I agree with that classification. I'd like to add some details though:

> [Download rules] [...] the inputs are explicitly specified, there are no implicit inputs from the system [...] These assumptions match the use case of language package managers such as maven and yarn.

Ideally that's the case. In practice this fails under certain circumstances. Language specific package managers like pip or yarn often make it hard to properly separate package download from package build or installation. If package build or installation ends up ocurring during repository rule evaluation, i.e. during fetch, then it is easy to incur a reference to system state. E.g. packages with native components often invoke whatever C/C++ compiler they find in PATH to build native components. Since repository rules don't have access to Bazel toolchains, yet, since they run in an earlier phase, these builds are usually not hermetic.
I'm mentioning this to point out that this ideal is not so easy to achieve with standard tools and its one of the areas where Nix can help, either by providing Nix built packages instead of invoking the package manager, or by providing a controlled toolchain through the environment or appropriate repository rule attributes.

> [Nix / as a system package manager] When used as a system package manager, for example [NixOS, Nix shell ...] remote execution can be supported easily, as the assumptions match other system package managers. The remote execution environment can either be a NixOS environment configured to match the host, or a container image configured to match a host [...]

It's worth pointing that when Nix is used in this way one loses out on the granularity that rules_nixpkgs provides. With rules_nixkpgs provided tools and toolchains Bazel can track which targets are affected by a change to an individual Nix package and can invalidate and rebuild only those targets. When Nix is used to generate an OS image (or similar) then Bazel does not have that information (or at least it depends very much on how each tool and toolchain is defined) and, to be sure, any change requires a clean rebuild.

Also, one of the benefits Nix provides is the ability to define the tools and their versions in code right in your repository. That means that PRs can contain changes to the Nix provided environment as well as to regular code. A remote execution setup needs to take this into account and must support multiple simultaneously existing configurations. In particular, one PR that updates a Nix package must not cause the entire remote execution cluster to be restarted with that new image, because then builds on other branches would run against the wrong versions. This is in contrast to some remote execution setups, where the remote executor image is centrally defined and updated in bulk. There are ways around this, but it's important to evaluate their costs carefully. For example, defining a Docker image to run each build action in can be a workaround. But, when evaluating that approach in practice we found that it can incur a significant overhead such that some types of build actions no longer benefit from remote execution.

> [Nix / as a language package manager] [...] for example when using nixpkgs_package to provide self contained outputs such as container image tarballs, remote execution is supported, as the files can be copied to remote executors and consumed there.

I agree with the general point. But, I wouldn't call this a language package manager use-case. Perhaps "Nix as an artifact provider" is better name. Language packages provided by Nix often have runtime dependencies on the Nix store, e.g. dynamic libraries like `glibc`, or package collections like `PYTHONPATH`, and cannot be used on another machine without ensuring that their transitive closure is available.

> In a sense there's nothing too special about nix in terms of integrating it into bazel: if one were to implement a rules_apt rule set that installs apt packages on demand via repository rules, it would be hitting all these issues as well. The special thing about nix is that using it on demand is so common that not being able to do so in a certain context is a big limitation.

Correct, perhaps with the added property that Nix stores packages in Nix store paths that depend on their inputs. A minor version apt package update will not change paths and a discrepancy across machines may go unnoticed, whereas with Nix it wouldn't. To be sure, this is an advantage of Nix in terms of correctness.

@sluongng Thanks for detailing this.
> Include the digest of the tree above as part of exec_properties of Bazel action (example), telling RBE platform to select builders matching this requirement

IIUC this comes back to the granularity point. A global image loses out on the per package granularity provided by rules_nixpkgs.

In that respect @layus' work seems like a good direction as it maintains that granularity. The particular implementation so far uses Nix's remote execution protocol to communicate the required Nix packages to the remote side before the build. But, thinking more generally, that doesn't necessarily have to be the way it's done. We could also use some other scheme to register required packages with the remote side.

**Comment by olebedev** [comment]
> @sluongng Thanks for detailing this.
>> Include the digest of the tree above as part of exec_properties of Bazel action (example), telling RBE platform to select builders matching this requirement
>
> IIUC this comes back to the granularity point. A global image loses out on the per package granularity provided by rules_nixpkgs.

@aherrmann, maybe we can add attributes that are required to be present for a particular action, in the similar way, in `exec_properties` and make sure we built only them, not the whole attrset? Assuming Nix code base is relatively small for cheap frequent fetches.

> In that respect @layus' work seems like a good direction as it maintains that granularity. The particular implementation so far uses Nix's remote execution protocol to communicate the required Nix packages to the remote side before the build. But, thinking more generally, that doesn't necessarily have to be the way it's done. We could also use some other scheme to register required packages with the remote side.

I agree with that. This decouples the problem from Bazel and turns it into an infrastructure challenge, which seems solvable and can be optimised.

**Comment by aherrmann** [comment]
> maybe we can add attributes that are required to be present for a particular action, in the similar way, in `exec_properties` and make sure we built only them, not the whole attrset? Assuming Nix code base is relatively small for cheap frequent fetches.

@olebedev Yes, that comes back to the suggestion in https://github.com/tweag/rules_nixpkgs/issues/180#issuecomment-1329081409. As laid out there there the challenge is to capture transitive dependencies correctly. E.g. a `cc_library` built with a Nix provided CC toolchain may have a dependency on Nix store paths, e.g. a shared library dependency like `glibc`. A downstream target, e.g. a `cc_binary` or a `haskell_binary` that depends on that `cc_library` will transitively depend on that Nix store path as well. We probably don't want to require users to collect these transitive dependencies manually in `exec_properties`.

> A promising suggestion came from @illicitonion. Namely, mark Nix imported targets with special platform properties that define which Nix derivation is required to run the action. Then, add a feature to Bazel to accumulate platform properties across transitive dependencies such that the transitive closure of required Nix derivations is communicated to the remote side for each action. Finally, extend the remote executor to parse these platform properties and ensure that the required Nix derivations are instantiated before running the action.

---

> I agree with that. This decouples the problem from Bazel and turns it into an infrastructure challenge, which seems solvable and can be optimised.

Yes, exactly.

**Comment by Jonpez2** [comment]
Hello!  Revisiting this after a long diversion into other, less important matters :)

If we did want to take the 'special platform properties' route, who would we need to engage, and how would we move it forward?

Thank you!

**Comment by aherrmann** [comment]
Hi @Jonpez2, thanks for checking in!

On the Bazel side that would require a feature request for the collecting of the transitive platform properties and discussion with the Bazel team about the exact shape of this. As mentioned above there are some open questions about how to control when properties are forwarded and when not. Some targets produce outputs that no longer need transitive Nix store paths at build or runtime. Generally, distinguishing build and runtime dependencies is another challenge with this approach. On the remote executor side this requires implementation work on the particular remote execution platform used to parse these platform properties and act on them.

All that said, as mentioned above, after exploring and discussing different approaches, I think the approach that @layus worked on and presented at Bazel eXchange is the most promising concept. Some details may need fleshing out or revisiting, but conceptually it's a very elegant approach to say, if a build rule knows a Nix store path, then that means that it has a legitimate direct or transitive dependency on it. So, we track which Nix store paths are requested at the loading phase (in the current implementation through Nix remote builds, but it doesn't have to be that), and then make sure that the build nodes can access them (in the current implementation through a shared network file system, but it doesn't have to be that). Bazel, or the remote execution protocol, doesn't have to be touched.

**Comment by sluongng** [comment]
@aherrmann I think you are aiming for something with higher "purity" here.

My take on @Jonpez2 's question is: If we are simply aiming to meet the current functionality of supporting custom docker images like this

```
    exec_properties = {
        "OSFamily": "Linux",
        "container-image": "docker://gcr.io/YOUR:IMAGE",
    },
```

with something like

```
    exec_properties = {
        "OSFamily": "Linux",
        "nix-pkg-revision": "17.09",
        "nix-pkg-sha256": "aaaaa...", # Optional
        "nix-shell": "//:shell.nix"
    },
```

Then it's simply a matter of engaging with an RBE software vendor to implement support for this on the server side.

It will have the same tradeoffs with supporting custom Docker image today: the platform will create and store a new root dir using Nix, then set it as the base root dir (and environment variables) before each action executed on the remote executor.

> What's impure about this?

If you update the Nix platform config, Bazel will not know about it to invalidate Action Cache.
This is also a common pain point for custom Docker image user today, but having such a feature allow our customers to adopt Bazel a lot faster and benefit from RBE much earlier.

If we could agree on this being a reasonable short-term solution for the Nix ecosystem, then I would love to collab with somebody to implement this feature as part of BuildBuddy's RBE offering.

**Comment by Jonpez2** [comment]
I’m personally pretty axed to get this
 - correct (I.e. pure in both nix and bazel terms)
 - minimally invalidating
 - transparent w.r.t. local vs remote execution
 - minimally incremental in terms of infra (like a globally distributed
filesystem seems kinda hard!)

I know I’m asking for a lot :(

**Comment by Silic0nS0ldier** [comment]
> I’m personally pretty axed to get this
>  - correct (I.e. pure in both nix and bazel terms)
>  - minimally invalidating
>  - transparent w.r.t. local vs remote execution
>  - minimally incremental in terms of infra (like a globally distributed
> filesystem seems kinda hard!)
> 
> I know I’m asking for a lot :(
> https://github.com/tweag/rules_nixpkgs/issues/180#issuecomment-1470303265

This is something I've been ideating on for awhile now. I believe such an outcome is possible, but on the Bazel side there is a major capability gap that effectively locks you into 100% local or 100% remote for building, unless you have some mechanism in place to ensure local and remote stay perfectly in sync. ...or if you are willing to accept some concessions and extra work (potentially a lot) some of this can be achieved manually with the existing spawn strategy controls (I personally don't recommend this).

The capability gap is execution platforms being entirely disconnected from spawn strategies, which can lead to scenarios where `platform(...)` declarations registered as extra execution platforms (be it via `register_execution_platforms(...)` in `WORKSPACE` and `MODULE`, or `--extra_execution_platforms=...` via `.bazelrc` and CLI) are used to configure targets/actions that are run locally despite (potentially) being incompatible.

e.g. When you have a macOS host and Linux remote executors, Bazel may (_will_ if the remote is unreachable or otherwise disabled) allocate actions configured for the remote platform to the macOS host which will depending on the inputs may proceed to fail.

In a setup where remote is the only enabled spawn strategy, you can get around this via RBE service specific logic and `exec_properties`. e.g. with EngFlow defining a worker pool and adding the pool name in `exec_properties` so the schedular knowns where to route actions.

There is an [approved proposal](https://github.com/bazelbuild/proposals/blob/afa38929278b7c20870125d1d32394955967263f/README.md?plain=1#L72) to address this capability gap, which if implemented (no idea if it has been started or not) should suit multi-platform build scenarios (especially those involving Nix) much better. [Execution Platforms vs. Strategies](https://docs.google.com/document/d/1U9HzdDmtRnm244CaRM6JV-q2408mbNODAMewcGjnnbM)

**Comment by Jonpez2** [comment]
That's a very interesting doc.  When I first started thinking about this whole thing (back when I was young and naive), I thought that injecting into something like the spawn strategies would make most sense as then there might be some chance of code sharing between the local and remote strategies.  Then maybe we could use flakes and nix run to give us extraordinary awesomeness.
The one point that seems to be explicitly out of scope in that docs is the collection of data from dependencies for feeding into the spawn strategy, which I think is a likely requirement here?

**Comment by Silic0nS0ldier** [comment]
> The one point that seems to be explicitly out of scope in that docs is the collection of data from dependencies for feeding into the spawn strategy, which I think is a likely requirement here?

Appropriate constraints added to `target_compatible_with` plus appropriate dependency propagation should address this issue once platforms can influence the selected spawn strategy. As for how `rules_nixpkgs` would need to evolve to support such mechanisms... That is TBD.

As for the spawn strategy side of the problem, it sounds like other projects are currently ranking higher priority wise (fair enough). I may have a go at implementing the proposal.

Not really in scope for this issue, but if anyone else wants to take a crack here are the notes I made for where changes would be necessary to implement the original proposal iteration (discussion around it made tracking down relevant source easier, the approved proposal should have a lot of overlap).

* Get list of strategies from execution platform.
  https://github.com/bazelbuild/bazel/blob/f861499cfc5c59b3b16c249c775ed20eef476218/src/main/java/com/google/devtools/build/lib/exec/SpawnStrategyRegistry.java#L105
  ```java
  resourceOwner.getExecutionPlatform()/*.getSpawnStrategies()*/;
  ```
* Strategy selection, likely no changes necessary.
  https://github.com/bazelbuild/bazel/blob/f861499cfc5c59b3b16c249c775ed20eef476218/src/main/java/com/google/devtools/build/lib/exec/SpawnStrategyResolver.java#L64
* Add `getSpawnStrategies()` and related bits to `PlatformInfo` class.
  https://github.com/bazelbuild/bazel/blob/f861499cfc5c59b3b16c249c775ed20eef476218/src/main/java/com/google/devtools/build/lib/analysis/platform/PlatformInfo.java#L43
* Update `PlatformInfo` construction to include strategy information.
  https://github.com/bazelbuild/bazel/blob/f861499cfc5c59b3b16c249c775ed20eef476218/src/main/java/com/google/devtools/build/lib/rules/platform/Platform.java#L41
* Update starlark `platform` function to accept spawn strategy info.
  https://github.com/bazelbuild/bazel/blob/f861499cfc5c59b3b16c249c775ed20eef476218/src/main/java/com/google/devtools/build/lib/rules/platform/PlatformRule.java#L31

**Comment by Jonpez2** [comment]
I assume people saw this link from the bazel-discuss mailing list

https://docs.google.com/document/d/1LD4uj1LJZa98C9ix3LhHntSENZe5KE854FNL5pxGNB4/edit?usp=sharing

**Comment by aaronmondal** [comment]
@uri-canva Thanks for linking this issue!

I think the upcoming remote execution implementation in [`rules_ll`](https://github.com/eomii/rules_ll) might be relevant to this issue since it also interleaves Nix and Bazel. Instead of invoking Nix via Bazel, we invoke Bazel wrapped in a Nix environment. So nixpkgs dependencies come from a flake and make up the environment in which Bazel runs. Then we use remote execution like this:

1. Wrap Bazel in an environment that sets all its runtime dependencies to something from nixpkgs: https://github.com/eomii/rules_ll/blob/main/bazel-wrapper/default.nix
2. Create a container image that uses the exact same wrapped Bazel as input. This will cause nix store paths on local installations and in the container to be identical (as long as the system architecture is similar enough): https://github.com/eomii/rules_ll/blob/main/rbe/image.nix
3. Use that container to generate remote execution-compatible toolchains. This will cause the generated toolchains to only refer to nix store paths: https://github.com/eomii/rules_ll/blob/14c1f431ceccd5b715a5df27639f2ff8d7506cf2/rbe/default/cc/BUILD#L135-L148

See also https://ll.eomii.org/setup/remote_execution.

Note that this functionality is not yet officially released and currently only works with upstream rules_ll, which is even less stable than the already highly volatile releases lol :sweat_smile:

With this setup you can actually run the remote execution toolchain locally because a local nix shell will install the same exact tools as are in the RBE container. This means that you can share caches between systems with pretty much prefect cache hit rate. Pretty cool! So you can have a remote executor build something large (in our case LLVM) and consume artifacts on a laptop and continue locally building custom targets on the laptop :blush: 

A drawback is that this system requires you to regenerate the entire RBE container and all toolchains if you change a dependency. If the Bazel wrapper script changes in any way the RBE container will have a different hash and the RBE toolchain will be incompatible with the previous one and require a full cache rebuild.

We have a [tool](https://github.com/eomii/rules_ll/blob/main/devtools/rbe.nix) that kind of does this stuff automatically, but at the moment it's a bit limited and only does the bare minimum we need for a basic toolchain.

Another drawback is that the Bazel wrapper currently only supports the Clang 15 toolchains from nixpkgs and hardcodes some flags around it. It would be desirable to make this more flexible so that it works with arbitrary (or at least "more") toolchain configs.

**Comment by sluongng** [comment]
https://github.com/pdtpartners/nix-snapshotter/blob/main/docs/architecture.md is a very interesting project.
It lets you define custom nix-based layers in your OCI-compatible container image.

When RBE worker downloads the container image, using the snapshotter, it would be able to construct the nix-layer using contents from the host's nix-store.

So this is 1 step closer to the RBE goal. The remaining pain would be how to build/update + release the container image before each Bazel build if there are some nix packages that were updated. But I think that could be done with a Bazel wrapper? WDYT? 🤔

**Comment by fhilgers** [comment]
As I understand, the problem is that the runtime dependencies for a package installed via the nix are missing on the remote executor. Is this right?

If so, could rules_nixpkgs just implement some feature like guix relocatable packs (https://guix.gnu.org/manual/en/html_node/Invoking-guix-pack.html) or use an already existing project like https://github.com/NixOS/bundlers or https://github.com/DavHau/nix-portable?

Basically, as far as I am aware, these projects use either linux user namespaces, proot or fakechroot to simulate the nix store being in the usual location /nix/store while it is actually somewhere else like in the bazel cache from the repository rule. All of those probably only work on linux, but maybe there exists a similar project for Macos that I just dont know about.

**Comment by aherrmann** [comment]
> As I understand, the problem is that the runtime dependencies for a package installed via the nix are missing on the remote executor. Is this right?

Yes, that's right.

> could rules_nixpkgs just implement some feature like guix relocatable packs

For some specific use-cases that could work, such as executables used as build tools or runtime dependencies. E.g. we could import a relocatable version of `hello`, place the artifact into a Bazel controlled output, Bazel would treat it as any other source file, and forward it to the RE worker.

It gets trickier for other use-cases. E.g. imported shared libraries have `RUNPATH` entries pointing to absolute Nix store paths and since they are not themselves executable they cannot be wrapped with something that sets up namespaces or fakechroot or the like.

Toolchains are also tricky. Let's say we import a CC toolchain from Nix and manage to make it relocatable using namespaces or the like. Now we can invoke the C compiler in a Bazel build action and it will find references to absolute Nix store paths, e.g. headers or standard libraries, thanks to the namespaces setup. But, (assuming dynamic linking) it will produce a binary that references shared objects in the Nix store, e.g. glibc. So, to execute that resulting binary we now also need a relocatable wrapper that captures the transitive runtime dependencies. So, this approach is somewhat viral and invasive. And there is an issue of duplication and artifact size. If we import two separate tools from Nix, they will likely have some overlap in their runtime closure, typically at least libc. This will be duplicated in the relocatable bundles.

Nonetheless, it's certainly an interesting approach and could work for some specific use-cases. You might be interested to read [this blog post](https://www.hdlfactory.com/post/2024/04/20/nix-bazel-%EF%B8%8F/) by @filmil and follow the links to the example project and related discussions on GitHub.

**Comment by domenkozar** [comment]
Note that [Tvix at NixCon 2024](https://talks.nixcon.org/nixcon-2024/talk/FLPMKH/) will talk about how tvix.dev has FUSE layer support, so you could share those store paths across remote nodes and other machines.

**Comment by layus** [comment]
Cool !

Le 23 octobre 2024 12:06:52 GMT+02:00, "Domen Kožar" ***@***.***> a écrit :

**Comment by rorychatterton** [comment]
> Toolchains are also tricky. Let's say we import a CC toolchain from Nix and manage to make it relocatable using namespaces or the like. Now we can invoke the C compiler in a Bazel build action and it will find references to absolute Nix store paths, e.g. headers or standard libraries, thanks to the namespaces setup. But, (assuming dynamic linking) it will produce a binary that references shared objects in the Nix store, e.g. glibc. So, to execute that resulting binary we now also need a relocatable wrapper that captures the transitive runtime dependencies. So, this approach is somewhat viral and invasive. 

If there are hard-coded dynamic links to the nix store in the built binaries - doesn't this issue just get pushed downstream to the point of release, either through the patching of the final binaries? 

Wouldnt this issue still exist in the FUSE/NFS style solution? Or are you saying that it's less common for that dynamic link to leak into the final executable?

I understand for many workloads, the outputs could be run inside docker containers or a nix host itself, but I could imagine this is a problem for people who want to use nix + bazel to build for edge/iot type targets.

Would we still need to look towards patchelf/install_name_tool to rewrite dependencies to point to the bazel's binary store instead?
